<a href="https://colab.research.google.com/github/Vitela-debug/Colab_S1_Clase-de-progralaci-n_273522_TVB/blob/main/Actividad_Calificada_07_Datos_Duplicados_273522.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo: Tania Vitela Barraza** | |
| **Matrícula: 273522** | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [2]:
# Escribe tu código aquí
hoteles_first=hoteles.duplicated(keep='first').sum()
hoteles_last=hoteles.duplicated(keep='last').sum()
hoteles_false=hoteles.duplicated(keep=False).sum()
print('Número de filas duplicadas (keep="first"):', hoteles_first)
print('->Marca como duplicadas las filas (reservaciones) repetidas después de la primera aparición')
print('Número de filas duplicadas (keep="last"):', hoteles_last )
print('->Marca como duplicadas las filas (reservaciones) repetidas antes de la última aparición')
print('Número de filas duplicadas (keep=False):', hoteles_false )
print('->Marca como duplicadas todas las filas (reservaciones) que son parte de los grupos de repetidas')

Número de filas duplicadas (keep="first"): 31994
->Marca como duplicadas las filas (reservaciones) repetidas después de la primera aparición
Número de filas duplicadas (keep="last"): 31994
->Marca como duplicadas las filas (reservaciones) repetidas antes de la última aparición
Número de filas duplicadas (keep=False): 40165
->Marca como duplicadas todas las filas (reservaciones) que son parte de los grupos de repetidas


---
---
* El código del ejercicio 1.1 sirve para identificar y contar las filas duplicadas en el DataFrame hoteles, tomando en cuenta todas sus columnas. Con` keep='first'` se considera original la primera aparición y se cuentan las que se repiten después; con `keep='last'` se considera original la última aparición y se cuentan las anteriores. Y finalmente  con `keep=False` se marcan y cuentan todas las filas que están repetidas. Se utiliza `.sum()` para obtener el número total de filas marcadas como duplicadas en cada caso
---
---

### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [3]:
# Escribe tu código aquí
porcentaje = (hoteles_first / len(hoteles)) * 100 #tambien se puede otomar el last

# Grupos repetidos
grupos_repetidos = hoteles.value_counts()
grupos_repetidos = (grupos_repetidos > 1).sum() #si una misma información de un hotel aparece mas de una vez, esas  filas forman un solo grupo repetido.

print("Porcentaje de filas sobrantes:", porcentaje, "%")
print("Número de grupos repetidos:", grupos_repetidos)

Porcentaje de filas sobrantes: 26.797889270458164 %
Número de grupos repetidos: 17


---
---
* El código del ejercicio 1.2 calcula el porcentaje dividiendo los duplicados repetidos entre el total de filas del Dataset. Después, `value_counts()` cuenta cuántas veces aparece cada combinación de valores en todas las columnas; al usar `> 1`, identifica las combinaciones que aparecen más de una vez y .`sum()` cuenta cuántos grupos repetidos existen.
---
---


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [4]:
# Escribe tu código aquí
hoteles_duplicados = hoteles.copy()
hoteles_duplicados['es_duplicado'] = hoteles_duplicados.duplicated(keep='first') #para identificar duplicados

filas_duplicadas = hoteles_duplicados.groupby('hotel')['es_duplicado'].agg('sum') #se agrupan por hotel y se suman las true de si es duplicado
filas_hoteles = hoteles_duplicados.groupby('hotel')['es_duplicado'].agg('count') #se agrupan por hotel y se cuentan las filas
Porcentajes = (filas_duplicadas/ filas_hoteles) * 100 #se calcula el porcejtaje de cada grupo

print('Filas duplicadas por hotel')
print(filas_duplicadas)
print('Porcentajes')
print(Porcentajes)


Filas duplicadas por hotel
hotel
City Hotel      25902
Resort Hotel     6092
Name: es_duplicado, dtype: int64
Porcentajes
hotel
City Hotel      32.650952
Resort Hotel    15.207189
Name: es_duplicado, dtype: float64


---
---
* El código del ejercicio 1.3 sirve para encontrar las filas duplicadas de cada hotel y saber qué porcentaje representan del total de reservas. Primero hace una copia de los datos y marca cuáles filas están repetidas con `hoteles_duplicados.duplicated(keep='first')`, después agrupa la información por hotel y cuenta los duplicados con `hoteles_duplicados.groupby('hotel')['es_duplicado'].agg('sum')`. Luego cuenta todas las reservas de cada hotel con hoteles_duplicados.groupby('hotel')['es_duplicado'].agg('count') y calcula el porcentaje de duplicados por hotel.
---
---


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

El problema de duplicados es importante, ya que representa una parte considerable de las reservas:

* City Hotel: 25,902 filas duplicadas, que representan el 32.65% de sus reservas.
* Resort Hotel: 6,092 filas duplicadas, que representan el 15.21% de sus reservas.

Por lo tanto, no afecta por igual a ambos hoteles. En City Hotel el porcentaje de duplicados es aproximadamente 17.44 % mayor que en Resort Hotel. Lo que singinfica que City Hotel tiene una mayor cantidad de reservas duplicadas en proporcion comparado con Resort Hotel.


---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [5]:
# Escribe tu código aquí

#para elegir las columnas por ordenar primero se eliminaron las que generarias problemas o confuciones que
#harian dificil el ordenamiento como reservation_status, deposit_type, o market_segment luego con las
#columnas restantes se tomaron aquellas que van de lo más general a lo más específico, por eso hotel es el primero
#luego año. luego mes y asi sucesivamente. Al final se tomaron tambien aquellas columnas que pueden aportar datos por caracteristicas
#Esto facilita encontrar y revisar los registros duplicados.

Filas_duplicadas = hoteles[hoteles.duplicated(keep=False)].sort_values(by=['hotel', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month','lead_time', 'adults', 'country', 'adr'])

Filas_duplicadas[['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'adults', 'country', 'market_segment', 'deposit_type', 'adr', 'reservation_status']].head(12)






,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40574,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40580,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40588,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40562,City Hotel,117,2015,August,1,1,ITA,Offline TA/TO,No Deposit,72.5,Check-Out
40585,City Hotel,117,2015,August,1,1,ITA,Offline TA/TO,No Deposit,72.5,Check-Out
40563,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40564,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40566,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40567,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40568,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled


---
---
* El código del ejercicio 2.1 sirve para encontrar los registros duplicados usando `duplicated(keep=False)`, ordenarlos con `sort_values()` según características como hotel, fecha, adultos, país y precio para facilitar su comparación, y finalmente seleccionar las columnas relevantes y usar `head(12)` para mostrar solo las primeras 12 filas y revisar si realmente son copias.
---
---

### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [6]:
# Escribe tu código aquí
grupos=hoteles.value_counts(dropna=False) # Contar cuántas veces aparece cada combinación de filas
tamanos_grupos = grupos.value_counts().sort_index() # Contar cuántos grupos hay de cada tamaño
print("10 tamaños de grupo más frecuentes:")
print(tamanos_grupos.sort_values(ascending=False).head(10)) #repeticiones de las filas (grupo 1--- XXX filas tienen 1 copia)

tamano_maximo = grupos.max() #cantidad de repeticiones de las mismas filas
print("\nTamaño del grupo más grande:", tamano_maximo)

# Mostrar la combinación de valores del grupo más grande
print("\nCombinación del grupo más grande:")
print(grupos[grupos == tamano_maximo]) #para ver la info de la fila que mas se repite


10 tamaños de grupo más frecuentes:
count
1     79225
2      5034
3       969
4       497
5       275
6       171
10      123
7       115
8       103
20       97
Name: count, dtype: int64

Tamaño del grupo más grande: 180

Combinación del grupo más grande:
hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel  1            277        2016               November            46                        7                          1                        2                     2       

---
---
* El código del ejercicio 2.2 sirve para analizar las filas repetidas del DataFrame: primero, `value_counts(dropna=False)` cuenta cuántas veces aparece cada combinación completa de valores, incluyendo los datos faltantes; después, otro value_counts() cuenta cuántos grupos existen según su número de repeticiones y sort_index() los ordena por tamaño; con sort_values(ascending=False).head(10) se muestran los 10 tamaños de grupo más frecuentes; finalmente, max() encuentra cuántas veces se repite la combinación más repetida y grupos[grupos == tamano_maximo] muestra cuál es esa combinación de valores
---
---

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [7]:
# Escribe tu código aquí

duplicados = hoteles[hoteles.duplicated(keep=False)] #seleccionamos solo los duplicados incluyendo el primero

print("Market segment:")
print(pd.concat([ # para unir los dos resultados y axis=1 para hacerlo horizontal
    hoteles['market_segment'].value_counts(normalize=True).rename('Originales'),  #cuenta cuántas veces aparece cada categoría y normalize para que sea en proporcion
    duplicados['market_segment'].value_counts(normalize=True).rename('Duplicados')], axis=1)) #hacemos lo mismo pero solo para los duplicados


#se repite el proceso pero para la otra columna
print("\nDeposit type:")
print(pd.concat([
    hoteles['deposit_type'].value_counts(normalize=True).rename('Originales'),
    duplicados['deposit_type'].value_counts(normalize=True).rename('Duplicados')], axis=1))


#para la tasa de cancelacion calculamos el promedio porque los valores son 0 y 1 entonces el promedio representa directamente la proporcion de cancelaciones
print("\nTasa de cancelación:")
print("Dataset original:", hoteles['is_canceled'].mean())
print("Duplicados:", duplicados['is_canceled'].mean())



Market segment:
                Originales  Duplicados
market_segment                        
Online TA         0.473046    0.205876
Offline TA/TO     0.202856    0.303996
Groups            0.165935    0.416109
Direct            0.105587    0.034981
Corporate         0.044350    0.036699
Complementary     0.006223    0.001867
Aviation          0.001985    0.000473
Undefined         0.000017         NaN

Deposit type:
              Originales  Duplicados
deposit_type                        
No Deposit      0.876464    0.641678
Non Refund      0.122179    0.356430
Refundable      0.001357    0.001892

Tasa de cancelación:
Dataset original: 0.37041628277075134
Duplicados: 0.583866550479273


---
---
* El código del ejercicio 2.3 sirve para comparar las filas involucradas en duplicados con todo el dataset, con el objetivo de saber si los duplicados tienen características diferentes a las reservas en general. Primero, `duplicados = hoteles[hoteles.duplicated(keep=False)]` selecciona todas las filas que pertenecen a grupos repetidos, incluyendo la primera copia. Después, `value_counts(normalize=True)` calcula la proporción de cada categoría en `market_segment` y *deposit_type*, tanto para el dataset completo como para los duplicados; `rename()` les pone nombres para distinguirlos y `pd.concat(..., axis=1)` junta los resultados horizontalmente para compararlos. Finalmente, para` is_canceled` usamos` mean()` porque sus valores son 0 y 1, por lo que el promedio representa directamente la proporción de reservas canceladas. La comparación es relevante porque buscamos identificar si las filas duplicadas están distribuidas de manera diferente al dataset general, por ejemplo, si se concentran más en cierto segmento, tipo de depósito o tienen una tasa de cancelación mayor. Así podemos conocer las características y el posible impacto de los duplicados antes de decidir qué hacer con ellos.
---
---


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

a) No se puede asegurar que todas sean errores, porque no existe un identificador de reserva (como un ID). Sin embargo, hay evidencia fuerte de registros repetidos: existen 79225  grupos que aparecen duplicados, 5034 grupos que aparecen 2 veces, 969 que aparecen 3 veces y el grupo más grande aparece 180 veces exactamente igual (lo que representa un caincidencia poco probable recordando que tenemos fechas de registro y de recervacion como parte de las caracterirsticas). Además, los duplicados tienen una tasa de cancelación de 58.4%, frente a 37.0% en el dataset completo (lo que significa que dentro de las filas que están involucradas en duplicados, hay proporcionalmente más cancelaciones), y Groups representa 41.6% de los duplicados frente a 16.6% del total (es decir, los duplicados están más concentrados en reservas del segmento Groups). Esto sugiere que los duplicados no son una muestra aleatoria y podrían estar afectando el análisis.

b) duplicated() considera los NaN como parte de la combinación de valores al identificar filas repetidas. Esto es importante porque el grupo más grande, de 180 repeticiones, tiene agent = NaN y company = NaN. Por lo tanto, varias reservas distintas podrían coincidir en todos los datos disponibles simplemente porque esas columnas están vacías. Los NaN aumentan la incertidumbre sobre si una fila duplicada realmente representa la misma reserva. Por lo que no eliminaría inmediatamente las filas duplicadas pero si prestaría mayor atención a los grupos grandes como el del 180 repeticiones una vez verificando los datos si las eliminaría al comparar el resto de las posibles respuestas en las columnas de agent y company dado que es poco probable que tengan variables mayores a grupos grandes como 180.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [8]:
# Escribe tu código aquí
hoteles_limpio = hoteles.drop_duplicates(keep='first').reset_index(drop=True) ##para identificar duplicados y mantener solo el primero, luego reset_index renumera y (drop=True) para no conservar el indice anterior índice anterior
print('Filas antes:', len(hoteles)) #filas del original
print('Filas después:', len(hoteles_limpio)) #filas sin duplicados exactos
print('Filas eliminadas:', len(hoteles) - len(hoteles_limpio)) #diferencia

Filas antes: 119390
Filas después: 87396
Filas eliminadas: 31994


---
---
* El código del ejercicio 3.1 elimina los registros duplicados de hoteles, conservando el primero con keep='first'. Después, reset_index(drop=True) vuelve a numerar las filas desde 0 sin conservar el índice anterior. Finalmente, len() cuenta las filas antes y después de la limpieza, y la resta muestra cuántas filas duplicadas fueron eliminadas.
---
---

### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [9]:
# Escribe tu código aquí
tamaño_original = hoteles.groupby('hotel').size() #Agrupamiento por hoteles y ver el tamaño
medias_original = hoteles.groupby('hotel')[['is_canceled', 'adr']].mean() #sacar las medias de las dos columnas por grupos
resumen_original = pd.concat([tamaño_original.rename('Tamaño'),medias_original], axis=1).reset_index() #juntar la info en un solo Dataframe y inciar de nuevo a numerar

#Mismo proceso pero para hoteles_limpio
tamaño_limpio = hoteles_limpio.groupby('hotel').size()
medias_limpio = hoteles_limpio.groupby('hotel')[['is_canceled', 'adr']].mean()
resumen_limpio = pd.concat([tamaño_limpio.rename('Tamaño'),medias_limpio], axis=1).reset_index()

print('resumen_original')
print(resumen_original)
print('resumen_limpio')
print(resumen_limpio)

resumen_original
          hotel  Tamaño  is_canceled         adr
0    City Hotel   79330     0.417270  105.304465
1  Resort Hotel   40060     0.277634   94.952930
resumen_limpio
          hotel  Tamaño  is_canceled         adr
0    City Hotel   53428     0.300386  110.985944
1  Resort Hotel   33968     0.234809   99.025346


---
---
* El código sirve para resumir y comparar la información de las reservas por tipo de hotel. Primero, `groupby('hotel')` agrupa los datos en City Hotel y Resort Hotel,  `.mean()` calcula el promedio de is_canceled y adr y mientras que `.size()` cuenta cuántas reservas hay en cada hotel. `rename()` cambia el nombre de la cantidad de reservas a Tamaño,` pd.concat(..., axis=1) `une los resultados en columnasen un solo DataFrame y `reset_index() ` convierte hotel nuevamente en una columna. En conjunto, obtenemos un resumen con el número de reservas, la tasa de cancelación y la tarifa diaria promedio de cada hotel para el caso de datos originales y con la limpieza.
---
---

### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [10]:
# Escribe tu código aquí
resumen_original['version'] = 'original' #Se agrega la columna de versión original
resumen_limpio['version'] = 'sin duplicados' #e agrega la columna de versión sin duplicados
comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True) #Se unen los df con pd.condat y  ignorar y crear nuevos indices
comparacion = comparacion.sort_values(by='hotel') #Se agrupan por hotel de nuevo para que queden juntos
comparacion.reset_index(drop=True, inplace=True) #Se vuelve a acomodar el indice
comparacion

,hotel,Tamaño,is_canceled,adr,version
0,City Hotel,79330,0.417270,105.304465,original
1,City Hotel,53428,0.300386,110.985944,sin duplicados
2,Resort Hotel,40060,0.277634,94.952930,original
3,Resort Hotel,33968,0.234809,99.025346,sin duplicados


---
---
* El código sirve para comparar el dataset original con el dataset limpio: primero agrega una columna para identificar cada versión, después` pd.concat()` une ambos DataFrames y `ignore_index=True` crea nuevos índices, `sort_values()` los ordena por hotel para facilitar la comparación y `reset_index(drop=True, inplace=True)` vuelve a acomodar los índices desde cero sin conservar los anteriores. En general permite ver cómo cambian las reservas, cancelaciones y tarifas antes y después de eliminar los duplicados.
---
---


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

a)  La métrica que más cambia es la tasa de cancelación (is_canceled-media) , especialmente en City Hotel. En City Hotel pasa de 41.7% a 30.0%, una disminución de aproximadamente 11.7 puntos porcentuales. En Resort Hotel pasa de 27.8% a 23.5%, una disminución de aproximadamente 4.3 puntos. Esto se relaciona con la Actividad 2, donde encontramos que las filas duplicadas tenían una tasa de cancelación de 58.4%, mucho mayor que el 37.0% del dataset completo. Por eso, al eliminar duplicados se reduce considerablemente la tasa de cancelación, sobre todo en City Hotel.

b) Los duplicados pueden hacer que el modelo dé más importancia a ciertos registros repetidos, provocando un aprendizaje sesgado. Además, como vimos que los duplicados tienen una tasa de cancelación mayor (58.4%) que el dataset completo (37.0%), podrían hacer que el modelo aprenda una relación incorrecta o exagerada entre ciertas características y la cancelación. También pueden causar fuga de información entre entrenamiento y evaluación si una misma reserva o un registro idéntico aparece en ambos conjuntos, haciendo que el modelo parezca tener un rendimiento mejor del que realmente tendría con datos nuevos.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [11]:
# Escribe tu código aquí

# Verificar que no quedan duplicados exactos de forma directa
print("Duplicados exactos:", hoteles_limpio.duplicated().sum())

# Verificar que no hay índices repetidos
print("Índices repetidos:", hoteles_limpio.index.duplicated().sum())

print("Número de filas:", len(hoteles_limpio))
# Verificar que el índice va de 0 a len(hoteles_limpio) - 1 mostrando los primeros y ultimos indices
print(hoteles_limpio.index)



Duplicados exactos: 0
Índices repetidos: 0
Número de filas: 87396
RangeIndex(start=0, stop=87396, step=1)


---
---
* El código del ejercicio 4.1 comprueba que el DataFrame quedó correctamente limpio: duplicated().sum() verifica que no existan filas duplicadas, index.duplicated().sum() revisa que no haya índices repetidos, len() cuenta cuántas filas quedan y hoteles_limpio.index muestra los índices para comprobar que sean consecutivos desde 0 hasta la última fila (se compara con la cantidad de filas previamente calculadas).
---
---


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [12]:
# Escribe tu código aquí
columnas_rest = [col for col in hoteles_limpio.columns if col != "reservation_status_date"] #hacemos una lista de las columnas que se van a comparar mediante una lista de comprension, excluyendo con if
filas_duplicadas =hoteles_limpio[hoteles_limpio.duplicated(subset=columnas_rest, keep='first')] #seleccionamos solo los duplicados quitando el primero y de las columnas que previamente definimos
print("Filas sobrantes:", len(filas_duplicadas)) #Contabilizamos las filas del df

Col_mostrar = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'adr', 'reservation_status', 'reservation_status_date'] #seleccionamos las columnas que queremos mostrar
print(filas_duplicadas.sort_values(by=Col_mostrar)[Col_mostrar].head(10)) #mustras las primeras 10 filas ordenadas de acuerdo a las columnas solicitadas

print("\n Distribución de reservation_status")
print(filas_duplicadas['reservation_status'].value_counts()) #  para contar cuántas veces aparece cada valor dentro de la columna reservation_status


Filas sobrantes: 245
            hotel  lead_time  arrival_date_year arrival_date_month  \
35574  City Hotel          9               2015          September   
51268  City Hotel         12               2017               July   
54474  City Hotel         14               2015            October   
39432  City Hotel         14               2016              April   
40083  City Hotel         14               2016                May   
45797  City Hotel         15               2016           December   
48206  City Hotel         17               2017              April   
48546  City Hotel         18               2017              April   
39715  City Hotel         19               2016              April   
56789  City Hotel         21               2016            January   

       arrival_date_day_of_month     adr reservation_status  \
35574                         10   95.00           Canceled   
51268                         15  139.00           Canceled   
54474              

**Respuesta 4.2:**

¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

Las 245 filas podrían representar registros repetidos de reservas canceladas, donde la única diferencia entre los registros es reservation_status_date. No las eliminaría directamente, porque esta columna fue excluida intencionalmente al buscar duplicados y podría contener información relevante sobre cuándo se actualizó o canceló cada reserva (que puede servir para medir que tan frecuente son las cancelaciones). Primero analizaría el significado de esta fecha y comprobaría si las filas corresponden a registros históricos o a duplicados reales. Si se confirma que no aportan información adicional (se tienen la misma fecha en la columna excluida y se duplico por erro de captura o tipo de datos)  y que cada reserva debe aparecer una sola vez, entonces podrían eliminarse conservando un único registro.

---
---
* El código del ejercicio 4.2 utiliza `.columns` para obtener las columnas del DataFrame, una list comprehension con if para excluir `reservation_status_date`, .duplicated() para identificar las filas repetidas, `len() `para contar cuántas filas duplicadas existen, `.sort_values() `para ordenarlas, `.head(10)` para mostrar las primeras 10, `.value_counts()` para contar cuántas veces aparece cada valor de reservation_status, En general, el código permite identificar, contar y analizar las filas duplicadas del dataset excluyendo una de las columnas y verificando si las el estatus de las reservas, mostrando que es el mismo para todas las filas.
---
---

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

Para la limpieza se identificaron inicialmente los duplicados exactos utilizando todas las columnas del dataset. Se utilizó` drop_duplicates(keep='first')`, por lo que, cuando existían varias filas completamente iguales, se conservó la primera aparición y se eliminaron las siguientes. En este caso, la regla `keep='first'` no representa una diferencia significativa frente a `keep='last'`, ya que ambas alternativas identificaron 31,994 filas repetidas; sin embargo, se eligió first para mantener el primer registro encontrado.

El dataset original tenía 119,390 filas y, después de eliminar los duplicados exactos, quedaron 87,396 filas, por lo que se eliminaron 31,994 registros, equivalentes aproximadamente al 26.8% del dataset original. Después de la limpieza se verificó que no quedaran duplicados exactos ni índices repetidos.

Como análisis adicional, se revisaron posibles duplicados ignorando la columna reservation_status_date. Bajo este criterio se encontraron 245 filas sobrantes, todas con `reservation_status = "Canceled"`. Estas filas no fueron eliminadas, ya que `reservation_status_date` puede contener información relevante sobre la fecha asociada al cambio de estado de la reservación. Eliminarlas sin comprobar primero el significado de esta variable podría provocar pérdida de información.

Como supuesto, se considera que las filas completamente idénticas representan registros duplicados y que conservar una sola copia no afecta la información necesaria para el análisis. Una limitación es que no se cuenta con un **identificador único de reservación que permita confirmar de manera absoluta si dos registros corresponden a la misma reserva**. Además, los duplicados no exactos requieren un análisis adicional antes de eliminarlos, ya que algunas diferencias entre registros pueden representar información válida como es el caso de . `reservation_status_date` ya que una fecha puede significar información importnate para hacer predicciones de cancelaciones etc.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**


1. No convendría eliminarlos si cada fila representara un evento o registro histórico independiente de una misma reservación. En ese caso, aunque los datos sean exactamente iguales, podrían representar diferentes actualizaciones o movimientos y eliminarlos podría causar pérdida de información. Por ejemplo, si una misma reservación se registra dos veces porque ocurrió dos eventos diferentes en el sistema, aunque la información sea igual. Si eliminamos una fila, podríamos estar eliminando un registro válido. Otro ejemplo es si dos clientes podrían reservar el mismo hotel, para las mismas fechas, con el mismo número de personas, el mismo precio y todos los demas datos. Como el dataset no tiene un identificador único de reservación, no podemos saber si realmente es un duplicado o una reservación diferente.

2. Pediría principalmente un ID único de reservación, además de un ID de cliente y, si existe, una fecha/hora de creación o modificación del registro. Esto permitiría saber si dos filas corresponden realmente a la misma reservación o si son registros independientes, y conocer cuál es el registro más reciente.

3. duplicated().sum() == 0 no garantiza que un dataset esté libre de duplicados porque esta función busca filas que sean completamente iguales tomando en cuenta los valores de las columnas comparadas. Además, los datos pueden representar el mismo registro aunque alguna columna sea diferente. En nuestro análisis, por ejemplo, no quedaron duplicados exactos, pero encontramos 245 posibles duplicados al excluir reservation_status_date. Esto muestra que una diferencia en una columna, en este caso una fecha, puede hacer que pandas considere las filas como diferentes aunque puedan representar el mismo registro. También es importante considerar el tipo de dato de las columnas, ya que valores que representan lo mismo pueden estar almacenados de forma diferente, por ejemplo, 100 como número entero y 100.0 como decimal, o una fecha almacenada como texto y otra como tipo fecha. Por eso, antes de concluir que no existen duplicados, es necesario revisar tanto los valores como los tipos de datos y determinar qué columnas realmente identifican de manera única cada registro.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.